File: lib/common/dto/response_received_dto.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:netra_flutter/common/models/response.dart';

part 'response_received_dto.freezed.dart';
part 'response_received_dto.g.dart';

@freezed
abstract class ResponseReceivedDTO with _$ResponseReceivedDTO {
  const ResponseReceivedDTO._();

  const factory ResponseReceivedDTO({
    required Object? data,
    required int statusCode,
    required String? statusMessage,
    required Map<String, String?>? headers,
  }) = _ResponseReceivedDTO;

  factory ResponseReceivedDTO.fromJson(Map<String, dynamic> json) =>
      _$ResponseReceivedDTOFromJson(json);

  factory ResponseReceivedDTO.fromDataModel(ResponseReceived model) {
    return ResponseReceivedDTO(
      data: model.data,
      statusCode: model.statusCode,
      statusMessage: model.statusMessage,
      headers: model.headers,
    );
  }

  ResponseReceived toDataModel() {
    return ResponseReceived(
      data: data,
      statusCode: statusCode,
      statusMessage: statusMessage,
      headers: headers,
    );
  }
}

File: android/src/main/kotlin/com/example/netra_flutter/dto/ResponseReceivedDTO.kt
package com.example.netra_flutter.dto

import com.netra.library.NetraResponse

data class ResponseReceivedDTO(
    val data: Any?,
    val statusCode: Int,
    val statusMessage: String?,
    val isCache: Boolean?,
    val headers: Map<String, String>?,
) {
    companion object {
        fun<T> fromDataModel(response: NetraResponse.ResponseReceived<T>): ResponseReceivedDTO {
            return ResponseReceivedDTO(
                data = response.data,
                isCache = response.isCache,
                statusCode = response.statusCode,
                statusMessage = response.statusMessage,
                headers = response.headers,
            )
        }
    }

    fun toDataModel(): NetraResponse.ResponseReceived<*> {
        return NetraResponse.ResponseReceived(
            data = data,
            isCache = isCache,
            statusCode = statusCode,
            statusMessage = statusMessage,
            headers = headers,
        )
    }
}
